packages/fxa-auth-server/scripts/stripe-products-and-plans-to-firestore-documents/stripe-product-and-plans-converter.ts

```diff
@@ -0,0 +1,382 @@
+    productConfig.uiContent = this.uiContentMetadataToUiContentConfig(product);
+    productConfig.urls = this.urlMetadataToUrlConfig(product);
+
+    // Extended by ProductConfig
+    const { id, productSet, promotionCodes } = product.metadata;
+    productConfig.stripeProductId = id;
+    if (productSet) {
+      productConfig.productSet = productSet;
+    }
+    if (promotionCodes) {
+      productConfig.promotionCodes = commaSeparatedListToArray(promotionCodes);
+    }
+    return productConfig;
+  }
+
+  // TODO: JSDoc
+  // TODO: #12053: Improve heuristics
+  findLocaleStringFromStripePlan(plan: Stripe.Plan): null | string {
+    // Try to extract a locale from the plan nickname
+    const { nickname } = plan;
+    let locale = nickname
+      ?.split(' ')
+      .filter((w) => this.supportedLanguages.includes(w.toLowerCase()));
+    if (locale && locale.length > 0) {
+      return locale[0];
+    }
+    return null;
+  }
+
+  // Extract any locale data from a Stripe plan and convert it to ProductConfig.locales
+  stripePlanLocalesToProductConfigLocales(
+    plan: Stripe.Plan
+  ): ProductConfig['locales'] {
+    const locales: ProductConfig['locales'] = {};
+    const localeStr = this.findLocaleStringFromStripePlan(plan);
+    if (!localeStr) {
+      return locales;
+    }
+    const uiContent = this.uiContentMetadataToUiContentConfig(plan);
+    const urls = this.urlMetadataToUrlConfig(plan);
+    const support = this.supportMetadataToSupportConfig(plan);
+    locales[localeStr] = {
+      uiContent,
+      urls,
+      support,
+    };
+    return locales;
+  }
+
+  // TODO JSDoc
+  stripePlanToPlanConfig(
+    plan: Stripe.Plan,
+    productConfigId: string
+  ): PlanConfig {
+    // Firestore cannot serialize class instances
+    const planConfig: any = {};
+
+    planConfig.productConfigId = productConfigId;
+    planConfig.active = plan.active;
+
+    if (!plan.metadata) {
+      return planConfig;
+    }
+
+    const metadataKeys = Object.keys(plan.metadata);
+
+    // Optional BaseConfig
+    if (
+      metadataKeys.some((key) => key.toLowerCase().startsWith('capabilities'))
+    ) {
+      planConfig.capabilities =
+        this.capabilitiesMetadataToCapabilityConfig(plan);
+    }
+    if (metadataKeys.some((key) => key.toLowerCase().includes('url'))) {
+      planConfig.urls = this.urlMetadataToUrlConfig(plan);
+    }
+    if (
+      metadataKeys.some((key) => Object.keys(UiContentConfigKeys).includes(key))
+    ) {
+      planConfig.uiContent = this.uiContentMetadataToUiContentConfig(plan);
+    }
+    if (metadataKeys.some((key) => StyleConfigKeys.includes(key))) {
+      planConfig.styles = this.stylesMetadataToStyleConfig(plan);
+    }
+    if (metadataKeys.some((key) => key.toLowerCase().startsWith('support'))) {
+      planConfig.support = this.supportMetadataToSupportConfig(plan);
+    }
+    // TODO Do we need `locales` on the planConfig, since they'll exist on productConfig?
```

Comment:
@chenba : IIUC, we are moving the localized versions of product-specific strings from Stripe Plans into `productConfig.locales` – does this mean we should remove the currently optional `planConfig.locales` property? I omitted it for now in my `stripePlanToPlanConfig` method, but the optional key is still on the PlanConfig class and schema.

Currently, we use the localized strings off the Stripe plan if present. By moving this information into productConfig, we lose that association of a given plan with which locale to use, but I suppose we would detect the locale of the user on the client side and select from the list of `locales` in `productConfig`?

Reply:
Chatted with Barry sync; we do want a `locales` key on the `planConfig`, for example for the MDN team which is going to have tiered pricing -- they will want different strings for e.g. `product:details{x}` for the same locales, which would be overwritten in the current implementation of this `convert` method.

We agreed addressing this would be part of the follow-up ticket I previously filed to infer a locale from a Stripe Plan, #12053 .